TrustKit/Pinning/TSKSPKIHashCache.m

```diff
@@ -274,16 +274,25 @@ - (SecKeyRef)copyPublicKeyFromCertificate:(SecCertificateRef)certificate
     }
     
     // Get a public key reference for the certificate from the trust
-    SecTrustResultType result;
-    status = SecTrustEvaluate(trust, &result);
-    if (status != errSecSuccess)
+    // The certificate chain must be evaluated first in order to be able
+    // to determine which is the leaf certificate of the chain, and only
+    // then SecTrustCopyKey can be called
+    SecTrustResultType trustResult = 0;
+    NSError *error = NULL;
+    bool isChainTrusted = evaluateCertificateChainTrust(trust, &error);
+    (void)isChainTrusted; // Discard the chain trust result, not relevant for copying the public key
+    status = SecTrustGetTrustResult(trust, &trustResult);
+    if (error != NULL)
     {
-        TSKLog(@"Could not evaluate trust for the certificate, got status %d", status);
-        CFRelease(trust);
-        return nil;
+        if (trustResult != kSecTrustResultRecoverableTrustFailure)
```

Comment:
Last request/question: the previous code was not checking the actual `trustResult`, only the `status`. 

Hence, I don't think we need to check the `trustResult` here (other values than `kSecTrustResultRecoverableTrustFailure` are also acceptable, I think); only the presence of an error ?

Reply:
Not checking the trustResult at all leads to test failures, e.g. if I have:
```
    bool isChainTrusted = evaluateCertificateChainTrust(trust, &error);
    (void)isChainTrusted; // Discard the chain trust result, not relevant for copying the public key
    if (error != NULL)
    {
        TSKLog(@"Could not evaluate trust for the certificate: %@", [error localizedDescription]);
        CFRelease(trust);
        return nil;
    }
```
Then I see 30 test failures, among them:
```
Failing tests:
	-[TSKPinningValidatorTests testHandleChallengeCompletionHandlerPinningSuccessful]
	-[TSKPinningValidatorTests testVerifyAgainstAnyPublicKey]
	-[TSKPinningValidatorTests testVerifyAgainstBadPublicKeyPinningNotEnforced]
	-[TSKPinningValidatorTests testVerifyAgainstCAPublicKey]
	-[TSKPinningValidatorTests testVerifyAgainstInjectedCaPublicKey]
	-[TSKPinningValidatorTests testVerifyAgainstLeafPublicKey]
	-[TSKPinningValidatorTests testVerifyAgainstLeafPublicKeyAndBadPublicKey]
	-[TSKPublicKeyAlgorithmTests testExtractEcDsaSecp256r1]
	-[TSKPublicKeyAlgorithmTests testExtractRsa2048]
	-[TSKPublicKeyAlgorithmTests testExtractRsa4096]
```
My understanding is that `SecTrustEvaluateWithError()` will populate the error if the cert isn't trusted, even when the evaluation was successful (in the technical sense). So we need information from `SecTrustGetTrustResult()` in order to understand if it's a trust failure, or a technical failure. The following also works:
```
    SecTrustResultType trustResult = 0;
    NSError *error = NULL;
    bool isChainTrusted = evaluateCertificateChainTrust(trust, &error);
    (void)isChainTrusted; // Discard the chain trust result, not relevant for copying the public key
    status = SecTrustGetTrustResult(trust, &trustResult);
    if ((error != NULL) && (status != errSecSuccess))
    {
        TSKLog(@"Could not evaluate trust for the certificate: %@", [error localizedDescription]);
        CFRelease(trust);
        return nil;
    }
```
It doesn't look at the trust result directly, but it still calls `SecTrustGetTrustResult()` and evaluates the status. Is that preferrable? 
(trustResult is a mandatory argument, so unfortunately we cannot just ignore it)

Reply:
Done

Reply:
Thank you